wxMaxima: update to version 20.04.0 (April 2020)

MSoegtropIMC opened the pull request:
#### Description

This PR updates wxMaxima from Version 18.02 (February 2018) to 20.04.0 (April 2020). This is the result of a lengthy bug fixing session between me and the wxMaxima authors.

See: (https://github.com/wxMaxima-developers/wxmaxima/issues/1220)

###### Type(s)

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
macOS 10.15.4 19E287
(the command given to get the xcode version results in an error with latest Xcode)

Attention: this is known to not work (well) on macOS 10.15.0 ... 10.15.3 but it works on 10.15.4 and should work on 10.14 and before.
See for a general discussion of wxWidgets issues on 10.15.X:
http://trac.wxwidgets.org/ticket/18641
https://forums.wxwidgets.org/viewtopic.php?t=46362&p=195663
As far as I can tell these issues have been fixed in 10.15.4 - at least they magically went away after I updated from 10.15.3.
This might apply to other wxWidgets based applications as well.
I mentioned this in the notes field of the ports file, so that users are aware.

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? (I referenced a wxMaxima ticket, not a MacPorts ticket)
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`? (there are no automated tests - it is a GUI frontend)
- [x] tried a full install with `sudo port -vst install`? (See notes below)
- [x] tested basic functionality of all binary files? (also tested by 2 friends)

For install -vst I get the below warnings, which more look like issues in the trace system than in this ports file:
```
Warning: The following existing file was hidden from the build system by trace mode:
  /private/var/select/sh
Warning: The following existing files were hidden from the build system by trace mode:
  /opt
  /opt/local/bin/dot
  /opt/local/bin/lualatex
  /opt/local/bin/perl
  /private/var/select/sh
  /usr/X11R6
Warning: The following files inside the MacPorts prefix not installed by a port were accessed:
  /opt/local/bin/daemondo
  /opt/local/bin/port
  /opt/local/bin/port-tclsh
  /opt/local/bin/portf
  /opt/local/bin/portindex
  /opt/local/bin/portmirror
  /opt/local/include/.turd_MacPorts
  /opt/local/libexec/macports/bin/tclsh8.5
  /opt/local/man
  /opt/local/sbin/.turd_MacPorts
```

MSoegtropIMC commented:
P.S.: if you tell me how to put in my Github account name as maintainer (I don't want to put in an email), I am willing to take ownership of wxMaxima for OSX. I also have write access to the wxMaxima github repo.

MSoegtropIMC commented:
This is my first MacPorts PR and I am a bit lost with the CI feedback. On latest Xcode it works. For 9.X and 8.X I got:
```
Installing dependency (16 of 100) 'p5.28-locale-gettext' with variants '' (requesting '') ... [FAIL]
Installing dependency (16 of 100) 'p5.28-locale-gettext' with variants '' (requesting '') ... [FAIL]
```
and for 7.X I see only this error:
```
Building 'wxMaxima' ... [ERROR] maintainers: .
```
I would appreciate some advice what I should do about this.

MSoegtropIMC commented:
Is it common that PRs hang around for almost 2 weeks without any comment? This is not very encouraging. This is my first MacPorts PR and it was a lot of work to prepare it together with the wxMaxima maintainers. I have a lot of friends which are waiting for it and I need to explain them how to setup a local MacPorts repo until this is merged.

mf2k commented:
I don't think anyone feels knowledgeable enough to even review it. But I do have some comments for you. I will review shortly.

mf2k commented on math/wxMaxima/Portfile:
```diff
@@ -1,112 +1,78 @@
 license             GPL-2
-# openssl is a dependency of curl in cmake
-license_noconflict  openssl
+license_noconflict  openssl # openssl is a dependency of curl in cmake, so build only
```
Please keep the comment on the line preceding the `license_noconflict` line.

mf2k commented on math/wxMaxima/Portfile:
```diff
@@ -1,112 +1,78 @@
-
-    livecheck.type  none
-}
+homepage            https://github.com/wxMaxima-developers/wxmaxima
```
This `homepage` is set to this value by the `github.setup` and so this line should be removed.

MSoegtropIMC replied:
Sorry, a victim of the cmake port group trial. Done now.

mf2k commented on math/wxMaxima/Portfile:
```diff
@@ -1,112 +1,78 @@
+wxWidgets.use       wxWidgets-3.0
 
-depends_lib         port:${wxWidgets.port}
+depends_lib         port:${wxWidgets.port} port:libomp-devel
```
- Change `depends_lib` to `depends_lib-append` to not overwrite any portgroup dependencies. 
- We don't like to depend on `-devel` ports. Does the `libomp` port really not work here? 
- Add the `libomp` dependency on another line.

MSoegtropIMC replied:
Thanks for the detailed review - I appreciate it!

A question: is libomp-devel an unreleased build / master branch build of libomp or is it a package variant which includes besides the library also C the headers (as e.g. in many linux distros)? I am not 100% sure but I think I tried port:libomp and I had issues with missing header files, so I assumed it was a version without headers. Indeed I would not like to depend on something like a master branch build.

Btw.: even the libomp-devel version is many years old.

mf2k replied:
The former. `-devel` ports are simply intended for the latest development version. But they often don't get up to date with even that. 

In Macports, all ports are supposed to have the development headers. It would be considered a bug if they don't. You might have to open a Trac ticket against the libomp port. The good news is that it has a maintainer so it will hopefully get fixed.

MSoegtropIMC replied:
Indeed libomp works fine. I guess when I tested it I was missing one of the 10 extra cmake settings one needs when one uses libomp, tried libompdevel and then forgot to revert this. I changed it now.

mf2k commented on math/wxMaxima/Portfile:
```diff
@@ -1,112 +1,78 @@
+depends_lib         port:${wxWidgets.port} port:libomp-devel
 
-depends_build       path:bin/cmake:cmake
+depends_build       port:cmake
```
- Restore the original `depends_build` on `path:bin/cmake:cmake` so that `cmake-devel` will also be allowed. 
- Per best practices, change `depends_build` to `depends_build-append` since port groups are being used.

mf2k commented on math/wxMaxima/Portfile:
```diff
@@ -1,112 +1,78 @@
+depends_build       port:cmake
 
-depends_run         bin:maxima:maxima
+depends_run         port:maxima port:gnuplot
```
- Change `depends_run` to `depends_run-append`.
- Use a new line when adding `port:gnuplot` and sort the two dependencies alphabetically. So it should appear first.

mf2k commented on math/wxMaxima/Portfile:
```diff
@@ -1,112 +1,78 @@
+}
+
 build {
     system -W ${worksrcpath}       "mkdir ./build"
-    system -W ${worksrcpath}/build "cmake -DwxWidgets_CONFIG_EXECUTABLE:FILEPATH=${wxWidgets.wxconfig} .."
+    system -W ${worksrcpath}/build "cmake -DwxWidgets_CONFIG_EXECUTABLE:FILEPATH=${wxWidgets.wxconfig} -DUSE_CPPCHECK=YES -DUSE_OPENMP=YES -DCMAKE_CXX_FLAGS=-I${prefix}/include/libomp -DOpenMP_CXX_FLAGS='-Xpreprocessor -fopenmp' -DOpenMP_CXX_LIB_NAMES=libomp -DOpenMP_libomp_LIBRARY=${prefix}/lib/libomp/libomp.dylib .."
+    system -W ${worksrcpath}/build "cmake -LA ."
     system -W ${worksrcpath}/build "cmake --build ."
 }
```
Since it uses `cmake` to build, have you tried using the cmake 1.1 portgroup instead of trying to manually build here?

MSoegtropIMC replied:
No, I didn't try this. I guess the lengthy cmake configure phase command line could be given in some way? Do you have a reference project using the camke portgroup you can recommend - of cause I can search for one but I am not sure all have the same quality, so if you have one at hand you can recommend as good example, it might save us one iteration.

mf2k replied:
> Actually I did remove a variant for the wxWidgets version, which I thought would be more damaging to maintenance than helpful.

Oh I see that now. Sorry I missed it. So please adjust the commit message to mention removing the -devel sub-port.

mf2k replied:
You can get a list by typing looking at [this track ticket](https://trac.macports.org/ticket/47197#comment:78). Some of the ports there use the older cmake 1.0 portgroup however. We want to move every port to the 1.1 version. 

I would look at `libssh` as an example. It uses the cmake 1.1 portgroup and adds some custom cmake configure arguments.

MSoegtropIMC replied:
> Since it uses `cmake` to build, have you tried using the cmake 1.1 portgroup instead of trying to manually build here?

I tried it and it doesn't work for me. It might be useful for building libraries and console apps, but building application packages like wxMaxima is a different level of complexity. When I faithfully convert my settings to using the portgroup it does build successfully, but there is no application package in the build tree (searching for it with find) and destroot fails. So I will revert to what I have and leave it for the future to debug the cmake port group (I already spend a good week getting it to work as is without breaking wxMaxima CI and wxMaxima builds on other platforms).

mf2k replied:
Thank you for trying. I'm not familiar enough with cmake to offer suggestions and I don't have time to look into it myself anyway.

MSoegtropIMC commented:
> The GitHub comment mentions removing variants, but it actually only removes the -devel sub-port. So change the comment to reflect that.

Actually I did remove a variant for the wxWidgets version, which I thought would be more damaging to maintenance than helpful.

mf2k commented on math/wxMaxima/Portfile:
```diff
@@ -1,112 +1,83 @@
+wxWidgets.use       wxWidgets-3.0
 
-depends_lib         port:${wxWidgets.port}
+depends_lib-append  port:${wxWidgets.port}
+depends_lib-append  port:libomp
```
We normally use a single command with line continuations. Also I assume we still want to allow `libomp-devel`. So write it like this instead:

```
depends_lib-append port:${wxWidgets.port} \
                   lib:${prefix}/lib/libomp/libomp:libomp

```

mf2k commented on math/wxMaxima/Portfile:
```diff
@@ -1,112 +1,83 @@
+depends_build-append path:bin/cmake:cmake
 
-depends_run         bin:maxima:maxima
+depends_run-append  port:gnuplot
+depends_run-append  port:maxima
```
Same change here. Use line continuations instead.

MSoegtropIMC commented on math/wxMaxima/Portfile:
```diff
@@ -1,112 +1,83 @@
-    
-    livecheck.regex Version-(\[a-z0-9.\]+)${extract.suffix}
-}
+github.setup        wxMaxima-developers wxmaxima ${version} Version-
```
OK. I removed the long_description since it referenced version, but I guess this was anyway redundant.

MSoegtropIMC replied:
Hmm I found long_description is required (by port lint). What put people in there typically? How long should the long_description be?

MSoegtropIMC replied:
I added something as long_description which I think makes more sense than the `${description} Version : ${version}` it was originally.

mf2k replied:
I don't have a good description of the long description. You kind of just look at other ports for reference. Personally, I think it is the text mostly copied from the developers site. The (short) description then is a one line condensed version of that. The version generally should not be in either one.

MSoegtropIMC commented:
@mf2k: thanks for the quick review - I appreciate it, also that you take the time to explain subtle details to a MacPorts new comer.

One question: as I wrote I could take ownership of this, but would prefer to link it to my github account name rather than an email. Is there a way to do that?

mf2k commented:
To become maintainer, you add both. The [guide covers it well](https://guide.macports.org/#reference.keywords.maintainers).

mf2k commented on math/wxMaxima/Portfile:
```diff
@@ -1,112 +1,84 @@
 
-depends_run         bin:maxima:maxima
+depends_run-append  port:gnuplot \
+                    port:maxima
```
Can maxima-devel also be allowed? If so, you need to change it to a bin type dependency like for the cmake dependency.

MSoegtropIMC replied:
I think it is more likely that developers would have a maxima in home - maxima is mostly lisp and interpreted or built on demand. wxMaxima allows to choose a different maxima and maxima allows to choose a different lisp library folder. Developers know how to handle that. For the non-expert user I want to to have as few options as possible. For libomp and cmake  it makes sense because some people might need devel for something unrelated, but maxima and wxMaxima are more of a unit and it is not likely that some other packages requires maxima-devel as dependency.

mf2k replied:
No port should require a dependency on a `-devel` port. Since one cannot have both installed, it would be very inconvenient for someone wanting to test `maxima-devel`. By default, it will install `maxima` for any user. If `maxima-devel` can work, it is preferred to allow it with a bin (in this case) dependency. Also it was the previous setting and I don't think we should change that.

MSoegtropIMC replied:
Sorry, I am stubborn about this. My point is that:
- Maxima changes more frequently than wxMaxima does and saying "can work with -devel" means that I should actually test this. We are not talking about an arbitrary minimally used library but a very complex command line tool and its GUI front end. I won't state that anything works here without testing it.
- I don't see any purpose in it. If Maxima-devel would be GIT master (which it is not), this would be a different thing. Developers would have an easy way to ask users to try latest master to test if something is fixed. But I don't see any value in supporting a four month old arbitrary development snapshot.
- Developers generally need very specific versions and know how to install these.
- Features which have no use do more harm than good because they mislead

In summary it would be extra testing effort to put it in without giving any advantage I see. If there would be a "Maxima-git-master" I would support it (also making clear that there is obviously no guarantee that this works) because this would have a value.

> No port should require a dependency on a -devel port

Then I am inclined to remove the support for openmp-devel. What purpose does it serve to support this? Only developers need to test such things and they should know how to put in an arbitrary openmp as openmp plain. At least this is what I did when testing this and it was easy enough to setup a local MacPorts patch repo. The likelihood that an arbitrary development snapshot of openmp does serve anybody any purpose is really pretty slim. If someone has a good reason to test that specific version (s)he has 99.9% likely good reason to test a bunch of other versions as well and anyway needs to setup a local port repo. And the remaining 0.1% of an already small number of people rounded to nearest integer is very likely 0.

MSoegtropIMC replied:
P.S.: to give a concrete example let's look at openmp-devel and my wxMaxima port. The pure existence of this did lead to:

- I spent ~45 minutes trying if it would fix my issues, which it didn't (the issue was that build options where missing)
- I forgot to remove the -devel, so that you had to comment on this in the review and I had to fix it
- If your review wouldn't have been so careful this might have slipped, so that wxMaxime would have dependend on openmp-devel which might have lead to all sorts of trouble with people using openmp for other stuff (which is not likely since openmp and Maxima are mostly used by scientist doing computations)
- It is most likely that everybody else having issues with openmp on OSX has the same experience since openmp is very special (it requires support in the compiler which Xcode by default does not offer) and the -devel variant doesn't help with this.

So the pure existence of the -devel variant did lead to quite a bit of trouble and it is purely due to your good review that it did not lead to much more trouble. That's the reason why I am not a big fan of it.

mf2k replied:
Since you are becoming maintainer (thank you!) it is your prerogative.

MSoegtropIMC commented:
> To become maintainer, you add both. The [guide covers it well](https://guide.macports.org/#reference.keywords.maintainers). But it is OK to only use the GitHub handle.

Thanks, I added me as maintainer and also updated the commit message.

mf2k commented on math/wxMaxima/Portfile:
```diff
@@ -1,112 +1,84 @@
 name                wxMaxima
-version             18.02.0
-maintainers         nomaintainer
+maintainers         @MSoegtropIMC
```
Can `openmaintainer` be added here? This is preferred for non-Macports committers so that we can fix various issues more easily without waiting for approval.

MSoegtropIMC replied:
I thought about that but at least initially I would prefer to be asked before things are changed I am now responsible for. I am almost always responsive in a shorter timeframe than this PR took. In case there are issues, we can still change this. As you have seen I have some strong opinions about details which others might think are unimportant.

> Thank you for your patience!

I say thank you for your detailed reviews - I learned a lot!

MSoegtropIMC commented:
Many thanks for your help!

ryandesign commented:
@MSoegtropIMC Please sign in to Trac so that we can assign [this ticket](https://trac.macports.org/ticket/60453) to you.

ryandesign commented:
Since there were already [other tickets assigned to you](https://trac.macports.org/query?status=!closed&owner=MSoegtropIMC) I assigned that one to you as well. But Trac did not seem to know your GitHub handle (I'm guessing because you never logged in to Trac) which probably means Trac did not know your email address and did not notify you of any of these tickets. You should take a look at them all.
